File: dt241m-controller/internal/controller/persistence_test.go
package controller_test

import (
	"path/filepath"
	"strings"
	"testing"

	"github.com/jacobgad/dt241m-controller/internal/mqtt"
	"github.com/jacobgad/dt241m-controller/internal/store"
	"github.com/jacobgad/dt241m-controller/internal/testutil"
)

func openStore(t *testing.T, name string) *store.SQLite {
	t.Helper()
	s, err := store.Open(filepath.Join(t.TempDir(), name))
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { s.Close() })
	return s
}

func openStoreAt(t *testing.T, path string) *store.SQLite {
	t.Helper()
	s, err := store.Open(path)
	if err != nil {
		t.Fatal(err)
	}
	return s
}

func TestSQLiteStoreRoundTrip(t *testing.T) {
	s := openStore(t, "a.sqlite")
	if rows, _ := s.LoadAll(); len(rows) != 0 {
		t.Fatal("expected empty store")
	}
	h := newHarness(t, harnessOptions{store: s})
	rx(testutil.RxFixtureMAC, rxIP, h.net)
	h.discover(t)
	rows, err := s.LoadAll()
	if err != nil || len(rows) != 1 {
		t.Fatalf("rows %v err %v", rows, err)
	}
	r := rows[0]
	if r.MAC != testutil.RxFixtureMAC || r.ID != rxID || r.Role != "receiver" || r.IP != rxIP || r.Name != nil ||
		*r.ReportedName != "ER02_286CD6D8" || *r.ProductName != "ProAVRx ER01" || *r.Firmware != "1.13471.133" || *r.Channel != 2 || r.Online {
		t.Fatalf("record %+v", r)
	}
	if r.LastSeenAt == nil || !r.LastSeenAt.Equal(r.FirstSeenAt) {
		t.Fatal("timestamps wrong")
	}
}

func TestReopeningDatabaseIsIdempotent(t *testing.T) {
	path := filepath.Join(t.TempDir(), "reopen.sqlite")
	first := openStoreAt(t, path)
	first.Close()
	second := openStoreAt(t, path)
	defer second.Close()
	if rows, err := second.LoadAll(); err != nil || len(rows) != 0 {
		t.Fatalf("rows %v err %v", rows, err)
	}
}

func TestOfflineAdapterSurvivesRestart(t *testing.T) {
	path := filepath.Join(t.TempDir(), "restart.sqlite")
	net := testutil.NewNetwork()
	rx(testutil.RxFixtureMAC, rxIP, net)

	first := openStoreAt(t, path)
	before := newHarness(t, harnessOptions{store: first, net: net})
	if err := before.ctrl.Start(before.ctx); err != nil {
		t.Fatal(err)
	}
	before.ctrl.Stop(before.ctx)
	first.Close()

	net.Remove(rxIP)
	second := openStoreAt(t, path)
	defer second.Close()
	after := newHarness(t, harnessOptions{store: second, net: net})
	if err := after.ctrl.Start(after.ctx); err != nil {
		t.Fatal(err)
	}
	defer after.ctrl.Stop(after.ctx)

	a, ok := after.ctrl.Registry.Get(testutil.RxFixtureMAC)
	if !ok || a.Online || a.IP != rxIP || *a.Channel != 2 {
		t.Fatalf("adapter %+v ok=%v", a, ok)
	}
	configs := after.mqtt.DiscoveryConfigs()
	if _, ok := configs["homeassistant/number/"+rxID+"/channel/config"]; !ok {
		t.Fatal("discovery not republished")
	}
	if after.mqtt.LastPayload(mqtt.ForDevice(testutil.RxFixtureMAC).Availability) != "offline" ||
		after.mqtt.LastPayload(mqtt.ControllerKnownState) != "1" || after.mqtt.LastPayload(mqtt.ControllerOnlineState) != "0" {
		t.Fatal("offline adapter not published correctly")
	}
	if rows, _ := second.LoadAll(); len(rows) != 1 {
		t.Fatal("adapter was deleted")
	}
	if len(net.AllWrites()) != 0 {
		t.Fatal("restart wrote to hardware")
	}
}

func TestRestartProbesLastKnownIPBeforeFullScan(t *testing.T) {
	s := openStore(t, "probe.sqlite")
	net := testutil.NewNetwork()
	rx(testutil.RxFixtureMAC, rxIP, net)
	before := newHarness(t, harnessOptions{store: s, net: net})
	_ = before.ctrl.Start(before.ctx)
	before.ctrl.Stop(before.ctx)

	net.ResetRequests()
	after := newHarness(t, harnessOptions{store: s, net: net})
	_ = after.ctrl.Start(after.ctx)
	defer after.ctrl.Stop(after.ctx)

	if reqs := net.Requests(); len(reqs) == 0 || reqs[0].IP != rxIP {
		t.Fatal("last-known IP was not probed first")
	}
	a, _ := after.ctrl.Registry.Get(testutil.RxFixtureMAC)
	availability := after.mqtt.PayloadsOn(mqtt.ForDevice(testutil.RxFixtureMAC).Availability)
	if !a.Online || availability[0] != "offline" || availability[len(availability)-1] != "online" {
		t.Fatalf("availability %v online=%v", availability, a.Online)
	}
}

func TestRestartDoesNotReapplyRoutes(t *testing.T) {
	s := openStore(t, "routes.sqlite")
	net := testutil.NewNetwork()
	d := rx(testutil.RxFixtureMAC, rxIP, net)
	before := newHarness(t, harnessOptions{store: s, net: net})
	_ = before.ctrl.Start(before.ctx)
	if _, err := before.ctrl.RequestChannelChange(before.ctx, testutil.RxFixtureMAC, 7); err != nil {
		t.Fatal(err)
	}
	before.ctrl.Stop(before.ctx)

	d.SetReported(1)
	after := newHarness(t, harnessOptions{store: s, net: net})
	_ = after.ctrl.Start(after.ctx)
	defer after.ctrl.Stop(after.ctx)
	if d.ReportedChannel() != 1 || len(net.WritesTo(rxIP)) != 1 || after.mqtt.LastPayload(mqtt.ForDevice(testutil.RxFixtureMAC).ChannelState) != "1" {
		t.Fatal("route was reapplied")
	}
}

func TestNameSurvivesRestart(t *testing.T) {
	s := openStore(t, "name.sqlite")
	net := testutil.NewNetwork()
	rx(testutil.RxFixtureMAC, rxIP, net)
	before := newHarness(t, harnessOptions{store: s, net: net})
	_ = before.ctrl.Start(before.ctx)
	before.ctrl.Rename(before.ctx, testutil.RxFixtureMAC, "Main Projector")
	before.ctrl.Stop(before.ctx)

	after := newHarness(t, harnessOptions{store: s, net: net})
	_ = after.ctrl.Start(after.ctx)
	defer after.ctrl.Stop(after.ctx)
	a, _ := after.ctrl.Registry.Get(testutil.RxFixtureMAC)
	if a.Name == nil || *a.Name != "Main Projector" || *a.ReportedName != "ER02_286CD6D8" {
		t.Fatalf("adapter %+v", a)
	}
	if after.mqtt.LastPayload(mqtt.ForDevice(testutil.RxFixtureMAC).NameState) != "Main Projector" {
		t.Fatal("name state wrong")
	}
	cfg := after.mqtt.DiscoveryConfigs()["homeassistant/number/"+rxID+"/channel/config"]
	if cfg["device"].(map[string]any)["name"] != "Main Projector" {
		t.Fatal("device name not applied")
	}
}

func TestDHCPMovePersistsOneRecordWithSameNameAndIdentity(t *testing.T) {
	s := openStore(t, "dhcp.sqlite")
	net := testutil.NewNetwork()
	rx(testutil.RxFixtureMAC, "192.168.1.2", net)
	h := newHarness(t, harnessOptions{store: s, net: net})
	_ = h.ctrl.Start(h.ctx)
	defer h.ctrl.Stop(h.ctx)
	h.ctrl.Rename(h.ctx, testutil.RxFixtureMAC, "Main Projector")
	before := h.mqtt.DiscoveryConfigs()

	net.Move("192.168.1.2", "192.168.1.8")
	h.ctrl.RunDiscovery(h.ctx, "dhcp")

	rows, _ := s.LoadAll()
	if len(rows) != 1 || rows[0].IP != "192.168.1.8" || *rows[0].Name != "Main Projector" {
		t.Fatalf("rows %+v", rows)
	}
	after := h.mqtt.DiscoveryConfigs()
	if len(after) != len(before) {
		t.Fatal("discovery topics changed")
	}
	count := 0
	for topic, cfg := range after {
		if !strings.Contains(topic, rxID) {
			continue
		}
		count++
		device := cfg["device"].(map[string]any)
		if device["identifiers"].([]any)[0] != "dt241m:fc19286cd6d8" || device["name"] != "Main Projector" {
			t.Fatalf("device %v", device)
		}
	}
	if count != 4 || after["homeassistant/number/"+rxID+"/channel/config"]["unique_id"] != rxID+"_channel" {
		t.Fatal("identity changed")
	}
}

func TestDiscoveryPreservesName(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	d := rx(testutil.RxFixtureMAC, rxIP, h.net)
	h.discover(t)
	h.ctrl.Rename(h.ctx, testutil.RxFixtureMAC, "Main Projector")
	d.SetTemplate("dev_name", "ER02_RENAMED_ON_DEVICE")
	d.SetTemplate("version", "1.13471.999")
	h.discover(t)
	a, _ := h.ctrl.Registry.Get(testutil.RxFixtureMAC)
	if *a.Name != "Main Projector" || *a.ReportedName != "ER02_RENAMED_ON_DEVICE" || *a.Firmware != "1.13471.999" {
		t.Fatalf("adapter %+v", a)
	}
	rows, _ := h.store.LoadAll()
	if *rows[0].Name != "Main Projector" || *rows[0].ReportedName != "ER02_RENAMED_ON_DEVICE" {
		t.Fatal("store wrong")
	}
	if h.mqtt.LastPayload(mqtt.ForDevice(testutil.RxFixtureMAC).NameState) != "Main Projector" {
		t.Fatal("name state wrong")
	}
}

func TestRenameKeepsIdentity(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	rx(testutil.RxFixtureMAC, rxIP, h.net)
	h.discover(t)
	h.ctrl.Rename(h.ctx, testutil.RxFixtureMAC, "Main Projector")
	before := h.mqtt.DiscoveryConfigs()
	h.mqtt.Clear()
	h.net.ResetRequests()
	h.ctrl.Rename(h.ctx, testutil.RxFixtureMAC, "Auditorium Projector")
	after := h.mqtt.DiscoveryConfigs()
	if len(after) != 4 {
		t.Fatalf("expected 4 republished configs, got %d", len(after))
	}
	for topic, cfg := range after {
		prev := before[topic]
		if cfg["unique_id"] != prev["unique_id"] || cfg["state_topic"] != prev["state_topic"] || cfg["command_topic"] != prev["command_topic"] {
			t.Fatalf("identity changed on %s", topic)
		}
		if cfg["device"].(map[string]any)["name"] != "Auditorium Projector" {
			t.Fatal("device name not updated")
		}
	}
	for _, p := range h.mqtt.Published() {
		if strings.Contains(p.Topic, "Projector") {
			t.Fatal("name leaked into topic")
		}
	}
	if len(h.net.Requests()) != 0 {
		t.Fatal("rename touched hardware")
	}
	rows, _ := h.store.LoadAll()
	if rows[0].MAC != testutil.RxFixtureMAC || *rows[0].Name != "Auditorium Projector" {
		t.Fatal("store wrong")
	}
}

func TestNameValidationAndClearing(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	rx(testutil.RxFixtureMAC, rxIP, h.net)
	h.discover(t)
	nameTopic := mqtt.ForDevice(testutil.RxFixtureMAC).NameState

	h.ctrl.Rename(h.ctx, testutil.RxFixtureMAC, "  Lobby TV  ")
	a, _ := h.ctrl.Registry.Get(testutil.RxFixtureMAC)
	if *a.Name != "Lobby TV" {
		t.Fatal("not trimmed")
	}
	if o := h.ctrl.Rename(h.ctx, testutil.RxFixtureMAC, strings.Repeat("x", 65)); o.Renamed {
		t.Fatal("accepted too long")
	}
	if o := h.ctrl.Rename(h.ctx, testutil.RxFixtureMAC, "bad\x07name"); o.Renamed {
		t.Fatal("accepted control chars")
	}
	if o := h.ctrl.Rename(h.ctx, "00:11:22:33:44:55", "Ghost"); o.Renamed {
		t.Fatal("renamed unknown device")
	}
	if h.mqtt.LastPayload(nameTopic) != "Lobby TV" {
		t.Fatal("rejected rename changed state")
	}

	if o := h.ctrl.Rename(h.ctx, testutil.RxFixtureMAC, "   "); !o.Renamed || o.Name != nil {
		t.Fatalf("clear outcome %+v", o)
	}
	a, _ = h.ctrl.Registry.Get(testutil.RxFixtureMAC)
	if a.Name != nil || h.mqtt.LastPayload(nameTopic) != "ER02_286CD6D8" {
		t.Fatal("name not cleared")
	}
	rows, _ := h.store.LoadAll()
	if rows[0].Name != nil {
		t.Fatal("store not cleared")
	}

	h.mqtt.Deliver(mqtt.ForDevice(testutil.RxFixtureMAC).NameSet, "Stage Monitor")
	eventually(t, func() bool { return h.mqtt.LastPayload(nameTopic) == "Stage Monitor" }, "mqtt rename")

	cfg := h.mqtt.DiscoveryConfigs()["homeassistant/text/"+rxID+"/name/config"]
	avail := cfg["availability"].([]any)
	if len(avail) != 1 || avail[0].(map[string]any)["topic"] != mqtt.ControllerAvailability || cfg["max"] != 64.0 {
		t.Fatalf("name entity config %v", cfg)
	}
}

File: dt241m-controller/src/tests/persistence.test.ts
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { SqliteAdapterStore } from "../db/adapter-store.js";
import { openDatabase, type OpenedDatabase } from "../db/database.js";
import { adapters } from "../db/schema.js";
import { deviceTopics } from "../mqtt/topics.js";
import { RX_FIXTURE_MAC } from "./helpers/fixtures.js";
import { createHarness, type Harness } from "./helpers/harness.js";
import { SimulatedDevice, SimulatedNetwork } from "./helpers/simulated-network.js";

const RX_IP = "192.168.1.5";
const RX_ID = "dt241m_fc19286cd6d8";
const RX_TOPICS = deviceTopics(RX_FIXTURE_MAC);

let tempDir: string;
let opened: OpenedDatabase[] = [];

beforeEach(() => {
  tempDir = mkdtempSync(path.join(tmpdir(), "dt241m-test-"));
  opened = [];
});

afterEach(() => {
  for (const database of opened) database.close();
  rmSync(tempDir, { recursive: true, force: true });
});

function openStore(file = "adapters.sqlite") {
  const database = openDatabase(path.join(tempDir, file));
  opened.push(database);
  return { database, store: new SqliteAdapterStore(database.db) };
}

async function bootWithDevice(store: SqliteAdapterStore, network = new SimulatedNetwork()): Promise<Harness> {
  const h = createHarness({ store, network });
  await h.controller.start();
  return h;
}

describe("SQLite adapter store", () => {
  it("runs migrations and starts empty", () => {
    const { store, database } = openStore();
    expect(store.loadAll()).toEqual([]);
    expect(database.db.select().from(adapters).all()).toEqual([]);
  });

  it("reopening an existing database file is idempotent", () => {
    const first = openStore("reopen.sqlite");
    first.database.close();
    const second = openStore("reopen.sqlite");
    expect(second.store.loadAll()).toEqual([]);
  });

  it("round-trips an adapter through save and loadAll", async () => {
    const { store } = openStore();
    const network = new SimulatedNetwork();
    network.place(RX_IP, new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    const h = createHarness({ store, network });
    await h.controller.runDiscovery("seed");

    const [record] = store.loadAll();
    expect(record).toMatchObject({
      mac: RX_FIXTURE_MAC,
      id: RX_ID,
      role: "receiver",
      ip: RX_IP,
      name: null,
      reportedName: "ER02_286CD6D8",
      productName: "ProAVRx ER01",
      model: "am_8270_proavrx-eth_er01-pway-dt241",
      firmware: "1.13471.133",
      channel: 2,
      online: false
    });
    expect(typeof record!.firstSeenAt).toBe("number");
    expect(record!.lastSeenAt).toBe(record!.firstSeenAt);
  });
});

describe("restart behaviour", () => {
  it("keeps an offline adapter visible as unavailable after restart", async () => {
    const network = new SimulatedNetwork();
    network.place(RX_IP, new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    const first = openStore();
    const before = await bootWithDevice(first.store, network);
    await before.controller.stop();
    first.database.close();

    network.remove(RX_IP);
    const second = openStore();
    const after = createHarness({ store: second.store, network });
    await after.controller.start();

    const adapter = after.controller.registry.get(RX_FIXTURE_MAC);
    expect(adapter).toBeDefined();
    expect(adapter!.online).toBe(false);
    expect(adapter!.ip).toBe(RX_IP);
    expect(adapter!.channel).toBe(2);

    const configs = after.mqtt.discoveryConfigs();
    expect(configs.has(`homeassistant/number/${RX_ID}/channel/config`)).toBe(true);
    expect(configs.has(`homeassistant/text/${RX_ID}/name/config`)).toBe(true);
    expect(after.mqtt.lastOn(RX_TOPICS.availability)?.payload).toBe("offline");
    expect(after.mqtt.lastOn("dt241m/controller/known_devices/state")?.payload).toBe("1");
    expect(after.mqtt.lastOn("dt241m/controller/online_devices/state")?.payload).toBe("0");

    expect(second.store.loadAll()).toHaveLength(1);
    expect(network.writesTo(RX_IP)).toHaveLength(0);
    await after.controller.stop();
  });

  it("brings a persisted adapter online by probing its last-known IP before the full scan", async () => {
    const network = new SimulatedNetwork();
    network.place(RX_IP, new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    const first = openStore();
    const before = await bootWithDevice(first.store, network);
    await before.controller.stop();

    const after = createHarness({ store: first.store, network });
    network.requests.length = 0;
    await after.controller.start();

    expect(network.requests[0]?.ip).toBe(RX_IP);
    expect(after.controller.registry.get(RX_FIXTURE_MAC)?.online).toBe(true);
    const availability = after.mqtt.messagesOn(RX_TOPICS.availability).map((m) => m.payload);
    expect(availability[0]).toBe("offline");
    expect(availability[availability.length - 1]).toBe("online");
    expect(network.writesTo(RX_IP)).toHaveLength(0);
    await after.controller.stop();
  });

  it("does not reapply previous routes after restart", async () => {
    const network = new SimulatedNetwork();
    const device = network.place(RX_IP, new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    const { store } = openStore();
    const before = await bootWithDevice(store, network);
    await before.controller.requestChannelChange(RX_FIXTURE_MAC, 7);
    await before.controller.stop();

    device.reportedChannel = 1;
    const after = createHarness({ store, network });
    await after.controller.start();

    expect(device.reportedChannel).toBe(1);
    expect(network.writesTo(RX_IP)).toHaveLength(1);
    expect(after.mqtt.lastOn(RX_TOPICS.channelState)?.payload).toBe("1");
    await after.controller.stop();
  });

  it("name survives restart", async () => {
    const network = new SimulatedNetwork();
    network.place(RX_IP, new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    const { store } = openStore();
    const before = await bootWithDevice(store, network);
    await before.controller.rename(RX_FIXTURE_MAC, "Main Projector");
    await before.controller.stop();

    const after = createHarness({ store, network });
    await after.controller.start();
    const adapter = after.controller.registry.get(RX_FIXTURE_MAC)!;
    expect(adapter.name).toBe("Main Projector");
    expect(adapter.reportedName).toBe("ER02_286CD6D8");
    expect(after.mqtt.lastOn(RX_TOPICS.nameState)?.payload).toBe("Main Projector");
    const config = after.mqtt.discoveryConfigs().get(`homeassistant/number/${RX_ID}/channel/config`)!;
    expect((config["device"] as { name: string }).name).toBe("Main Projector");
    await after.controller.stop();
  });

  it("persists a DHCP move as one record with the same name and identity", async () => {
    const network = new SimulatedNetwork();
    network.place("192.168.1.2", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    const { store, database } = openStore();
    const h = await bootWithDevice(store, network);
    await h.controller.rename(RX_FIXTURE_MAC, "Main Projector");
    const configsBefore = h.mqtt.discoveryConfigs();

    network.move("192.168.1.2", "192.168.1.8");
    await h.controller.runDiscovery("dhcp");

    const rows = database.db.select().from(adapters).all();
    expect(rows).toHaveLength(1);
    expect(rows[0]).toMatchObject({ mac: RX_FIXTURE_MAC, lastKnownIp: "192.168.1.8", name: "Main Projector" });
    expect(h.mqtt.lastOn(RX_TOPICS.ipState)?.payload).toBe("192.168.1.8");

    const configsAfter = h.mqtt.discoveryConfigs();
    expect([...configsAfter.keys()].sort()).toEqual([...configsBefore.keys()].sort());
    const deviceConfigs = [...configsAfter.entries()].filter(([topic]) => topic.includes(RX_ID));
    expect(deviceConfigs.length).toBe(3);
    for (const [, payload] of deviceConfigs) {
      expect((payload["device"] as { identifiers: string[] }).identifiers).toEqual(["dt241m:fc19286cd6d8"]);
      expect((payload["device"] as { name: string }).name).toBe("Main Projector");
    }
    expect(configsAfter.get(`homeassistant/number/${RX_ID}/channel/config`)?.["unique_id"]).toBe(`${RX_ID}_channel`);
    await h.controller.stop();
  });
});

describe("names", () => {
  async function namedHarness() {
    const network = new SimulatedNetwork();
    const device = network.place(RX_IP, new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    const h = createHarness({ network });
    await h.controller.runDiscovery("seed");
    return { h, device };
  }

  it("discovery preserves the stored name even when hardware metadata changes", async () => {
    const { h, device } = await namedHarness();
    await h.controller.rename(RX_FIXTURE_MAC, "Main Projector");

    device.template["dev_name"] = "ER02_RENAMED_ON_DEVICE";
    device.template["version"] = "1.13471.999";
    await h.controller.runDiscovery("again");

    const adapter = h.controller.registry.get(RX_FIXTURE_MAC)!;
    expect(adapter.name).toBe("Main Projector");
    expect(adapter.reportedName).toBe("ER02_RENAMED_ON_DEVICE");
    expect(adapter.firmware).toBe("1.13471.999");
    const [stored] = h.store.loadAll();
    expect(stored?.name).toBe("Main Projector");
    expect(stored?.reportedName).toBe("ER02_RENAMED_ON_DEVICE");
    expect(h.mqtt.lastOn(RX_TOPICS.nameState)?.payload).toBe("Main Projector");
  });

  it("rename does not change MAC identity, topics, HA device identifier, or unique IDs", async () => {
    const { h } = await namedHarness();
    await h.controller.rename(RX_FIXTURE_MAC, "Main Projector");
    const before = h.mqtt.discoveryConfigs();
    h.mqtt.clear();

    await h.controller.rename(RX_FIXTURE_MAC, "Auditorium Projector");

    const after = h.mqtt.discoveryConfigs();
    expect([...after.keys()].sort()).toEqual([...before.keys()].sort());
    for (const [topic, payload] of after) {
      const previous = before.get(topic)!;
      expect(payload["unique_id"]).toBe(previous["unique_id"]);
      expect(payload["state_topic"]).toBe(previous["state_topic"]);
      expect(payload["command_topic"]).toBe(previous["command_topic"]);
      expect((payload["device"] as { identifiers: string[] }).identifiers).toEqual(["dt241m:fc19286cd6d8"]);
      expect((payload["device"] as { name: string }).name).toBe("Auditorium Projector");
    }
    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.mac).toBe(RX_FIXTURE_MAC);
    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.id).toBe(RX_ID);
    const [stored] = h.store.loadAll();
    expect(stored?.mac).toBe(RX_FIXTURE_MAC);
    expect(stored?.name).toBe("Auditorium Projector");
    expect(h.mqtt.lastOn(RX_TOPICS.nameState)?.payload).toBe("Auditorium Projector");
    for (const topic of h.mqtt.published.map((m) => m.topic)) {
      expect(topic).not.toMatch(/Projector/);
    }
  });

  it("clears the name with an empty string and falls back to reportedName", async () => {
    const { h } = await namedHarness();
    await h.controller.rename(RX_FIXTURE_MAC, "Main Projector");
    const outcome = await h.controller.rename(RX_FIXTURE_MAC, "   ");
    expect(outcome).toEqual({ status: "renamed", mac: RX_FIXTURE_MAC, name: null });
    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.name).toBeNull();
    expect(h.mqtt.lastOn(RX_TOPICS.nameState)?.payload).toBe("ER02_286CD6D8");
    const [stored] = h.store.loadAll();
    expect(stored?.name).toBeNull();
  });

  it("trims and validates names", async () => {
    const { h } = await namedHarness();
    await h.controller.rename(RX_FIXTURE_MAC, "  Lobby TV  ");
    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.name).toBe("Lobby TV");

    const tooLong = await h.controller.rename(RX_FIXTURE_MAC, "x".repeat(65));
    expect(tooLong.status).toBe("rejected");
    const controlChars = await h.controller.rename(RX_FIXTURE_MAC, "bad\u0007name");
    expect(controlChars.status).toBe("rejected");
    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.name).toBe("Lobby TV");
    expect(h.mqtt.lastOn(RX_TOPICS.nameState)?.payload).toBe("Lobby TV");

    const unknown = await h.controller.rename("00:11:22:33:44:55", "Ghost");
    expect(unknown.status).toBe("rejected");
  });

  it("renaming never writes to the DT241M hardware", async () => {
    const { h } = await namedHarness();
    h.network.requests.length = 0;
    await h.controller.rename(RX_FIXTURE_MAC, "Main Projector");
    expect(h.network.requests).toHaveLength(0);
  });

  it("accepts renames delivered over MQTT", async () => {
    const { h } = await namedHarness();
    h.mqtt.deliver(RX_TOPICS.nameSet, "Stage Monitor");
    await new Promise((resolve) => setTimeout(resolve, 5));
    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.name).toBe("Stage Monitor");
    expect(h.mqtt.lastOn(RX_TOPICS.nameState)?.payload).toBe("Stage Monitor");
  });

  it("the Name entity stays usable while the device is offline", async () => {
    const { h } = await namedHarness();
    const config = h.mqtt.discoveryConfigs().get(`homeassistant/text/${RX_ID}/name/config`)!;
    const availability = config["availability"] as Array<{ topic: string }>;
    expect(availability.map((a) => a.topic)).toEqual(["dt241m/controller/availability"]);
    expect(config["command_topic"]).toBe(RX_TOPICS.nameSet);
    expect(config["max"]).toBe(64);
  });
});
